src/polygons.jl

```diff
@@ -80,56 +85,95 @@ end
         else
-            push!(holes, ring)
+            push!(hole_inds, i)
         end
     end
-    for hole in holes
-        found = false
-        for i = 1:length(polygons)
-            if _inring(hole[1], polygons[i][1])
-                push!(polygons[i], hole)
-                found = true
-                break
+    if length(hole_inds) > 0
+        # Ignore Z dimension for extents, this is 2.5D
+        extents = map(r -> GI.extent(r)[(:X, :Y)], GI.getring(geom))
+        for h in hole_inds
+            hole = GI.getring(geom, h)
+            length(hole) > 0 || continue
+            hole_extent = extents[h]
+            found = false
+            for ic in indexcache
+                e = ic[1]
+                exterior = GI.getring(geom, e)
+                if Extents.intersects(hole_extent, extents[e]) && _inring(hole[1], exterior)
+                    push!(ic, h)
+                    found = true
+                    break
+                end
+            end
+            if !found
+                # The hole is not inside any ring, so make it a polygon.
+                # This is not intended behaviour but ESRI docs call it
+                # a "Dirty" ring. So it is a thing in the wild.
+                push!(indexcache, [h])
             end
-        end
-        if !found
-            # TODO: does this follow the spec? this should not happen with a correct file.
-            push!(polygons, SubPolygon([hole])) # hole is not inside any ring; make it a polygon
         end
     end
-    return polygons
+    return geom
 end
 
 function _inring(pt::AbstractPoint, ring::LinearRing)
-    intersect(i, j) =
-        (i.y >= pt.y) != (j.y >= pt.y) &&
-        (pt.x <= (j.x - i.x) * (pt.y - i.y) / (j.y - i.y) + i.x)
-    isinside = intersect(ring[1], ring[end])
+    length(ring) > 2 || return false
+    isinside = @inbounds _intersects(pt, ring[1], ring[end])
+    # This loop is 80% of Polygon read time
     for k = 2:length(ring)
-        isinside = intersect(ring[k], ring[k-1]) ? !isinside : isinside
+        intersects = @inbounds _intersects(pt, ring[k], ring[k-1])
+        isinside = intersects ? !isinside : isinside
     end
     return isinside
 end
 
 function _isclockwise(ring)
     clockwise_test = 0.0
-    for i in 1:GI.npoint(ring)-1
-        prev = ring[i]
-        cur = ring[i + 1]
+    for i in 1:length(ring)-1
+        prev = @inbounds ring[i]
+        cur = @inbounds ring[i + 1]
         clockwise_test += (cur.x - prev.x) * (cur.y + prev.y)
     end
     clockwise_test > 0
 end
 
+function _intersects(pt, i, j)
+    (i.y >= pt.y) != (j.y >= pt.y) &&
+    (pt.x <= (j.x - i.x) * (pt.y - i.y) / (j.y - i.y) + i.x)
+end
+
+# TODO add this to `Extents.union` for any `Tuple/AbstractArray` point
+function _union(extent::Extents.Extent, point::Tuple)
+    X = min(extent.X[1], point[1]), max(extent.X[2], point[1])
+    Y = min(extent.Y[1], point[2]), max(extent.Y[2], point[2])
+    return Extents.Extent(; X, Y)
+end
+
```

Comment:
want to move this to https://github.com/rafaqz/Extents.jl/pull/24 or maybe GeoInterface?

It would be cool to have `Extents.union(extent, geometry)`.

Reply:
Yeah I will, but not worth waiting for the release cycle for this PR

Reply:
We can add it for `AbstractArray` and `NamedTuple` as well

Reply:
Annoying we cant do it for all GeoInterface.jl points...